storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -500,17 +566,14 @@ VoteTracker getVote(UInt64 validatorIndex) {
     final AtomicReference<SlotAndBlockRoot> latestEpochBoundary = new AtomicReference<>();
     readLock.lock();
     try {
-      blockTree
-          .getHashTree()
-          .processHashesInChain(
-              blockRoot,
-              (root, parent) -> {
-                treeBuilder.childAndParentRoots(root, parent);
-                if (shouldPersistState(blockTree, root)) {
-                  latestEpochBoundary.compareAndExchange(
-                      null, new SlotAndBlockRoot(blockTree.getSlot(root), root));
-                }
-              });
+      blockMetadata.processHashesInChain(
+          blockRoot,
+          (root, slot, parent) -> {
+            treeBuilder.childAndParentRoots(root, parent);
+            if (shouldPersistState(slot, parent)) {
+              latestEpochBoundary.compareAndExchange(null, new SlotAndBlockRoot(slot, root));
```

Comment:
Doesn't this only save the first epoch boundary seen as processing hashes on the chain instead of saving the latest one?

Reply:
.compareAndExchange seems to only set the value to the second argument, if the reference is equal to the first value.

Reply:
We're walking from the head backwards through the chain and want to capture the one closest to head, which will be the first one seen.